File: jianzhioffer/thirdround/7.cpp
#include <cstddef>
#include <vector>
#include <unordered_map>
using namespace std;

struct TreeNode
{
        int val;
        TreeNode* left;
        TreeNode* right;
        TreeNode(int x): val(x), left(NULL), right(NULL) {}
};

class Solution
{
public: 
        TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder);

private:
        TreeNode* createRoot(vector<int>& preorder, int pre_left, int pre_right, vector<int>& inorder, int in_left, int in_right);
        unordered_map<int, int> map;
};

TreeNode* Solution::buildTree(vector<int>& preorder, vector<int>& inorder)
{
        if(preorder.empty() || inorder.empty() || (preorder.size()!=inorder.size())) return NULL;
        map = {}; 
        for(int i=0; i<inorder.size(); i++)
        {   
                map[inorder[i]] = i;
        }   
        TreeNode* root = createRoot(preorder, 0, preorder.size()-1, inorder, 0, inorder.size()-1);
        return root;
}

TreeNode* Solution::createRoot(vector<int>& preorder, int pre_left, int pre_right, vector<int>& inorder, int in_left, int in_right)
{
        if(pre_left>pre_right) return NULL;

        TreeNode* root = new TreeNode(preorder[pre_left]);
        if(pre_left==pre_right) return root;
        
        //根节点在中序序列的位置
        int in_root = map[preorder[pre_left]];

        //左子树长度
        int left_len = in_root-in_left;

        root->left = createRoot(preorder, pre_left+1, pre_left+left_len, inorder, in_left, in_root-1);
        root->right = createRoot(preorder, pre_left+left_len+1, pre_right, inorder, in_root+1, in_right);
        return root;
}


int main()
{
    vector<int> pre = {1,2};
    vector<int> in = {2,1};
    
    Solution obj;
    TreeNode* result = obj.buildTree(pre, in);
    return 0;
}

File: jianzhioffer/thirdround/7.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

//// 递归
//func buildTree(preorder []int, inorder []int) *TreeNode {
//	if len(preorder) != len(preorder) || len(preorder) == 0 {
//		return nil
//	} else if len(preorder) == 1 {
//		return &TreeNode{Val: preorder[0]}
//	}
//
//	m := make(map[int]int)
//	for i, v := range inorder {
//		m[v] = i
//	}
//	return buildSubTree(preorder, inorder, 0, 0, len(inorder)-1, m)
//}
//
//func buildSubTree(preorder []int, inorder []int, preIndex int, inLeftIndex int, inRightIndex int, m map[int]int) *TreeNode {
//	if preIndex >= len(preorder) || inLeftIndex > inRightIndex {
//		return nil
//	}
//	node := &TreeNode{Val: preorder[preIndex]}
//	inIndex := m[preorder[preIndex]] // 父节点在中序遍历序列的位置
//	node.Left = buildSubTree(preorder, inorder, preIndex+1, inLeftIndex, inIndex-1, m)
//	node.Right = buildSubTree(preorder, inorder, preIndex+(inIndex-inLeftIndex)+1, inIndex+1, inRightIndex, m)
//	return node
//}

// 递归，利用golang slice底层共享同一数组的特性
func buildTree(preorder []int, inorder []int) *TreeNode {
	if len(preorder) != len(preorder) || len(preorder) == 0 {
		return nil
	} else if len(preorder) == 1 {
		return &TreeNode{Val: preorder[0]}
	}

	idx := 0
	for ; idx < len(inorder); idx++ {
		if inorder[idx] == preorder[0] {
			break
		}
	}

	node := &TreeNode{preorder[0], nil, nil}
	// slice为左闭右开
	node.Left = buildTree(preorder[1:idx+1], inorder[:idx])
	node.Right = buildTree(preorder[idx+1:], inorder[idx+1:])
	return node
}

func main() {
	preorder := []int{3, 9, 20, 15, 7}
	inorder := []int{9, 3, 15, 20, 7}
	root := buildTree(preorder, inorder)
	print(root.Val)
}
